api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -842,3 +842,83 @@ message RequestIDExtension {
   // Request ID extension specific configuration.
   google.protobuf.Any typed_config = 1;
 }
+
+// If this is used, the cluster will only operate on one of the possible upstream protocols (HTTP/1.1, HTTP/2).
+// If :ref:`http2_protocol_options <envoy_api_field_config.cluster.v3.Cluster.http2_protocol_options>` are
+// present, HTTP2 will be used, otherwise HTTP1.1 will be used.
+message ExplicitHttpConfig {
+  oneof protocol_config {
+    config.core.v3.Http1ProtocolOptions http_protocol_options = 1;
+
+    config.core.v3.Http2ProtocolOptions http2_protocol_options = 2;
+  }
+}
+
+// If this is used, the cluster can use either of the configured protocols, and
+// will use whichever protocol was used by the downstream connection.
+message UseDownstreamHttpConfig {
+  config.core.v3.Http1ProtocolOptions http_protocol_options = 1;
+
+  config.core.v3.Http2ProtocolOptions http2_protocol_options = 2;
+}
+
+// HttpProtocolOptions specifies Http upstream protocol options. This object
+// is used in
+// :ref:`typed_extension_protocol_options<envoy_api_field_config.cluster.v3.Cluster.typed_extension_protocol_options>`,
+// keyed by the name `envoy.filters.network.http_connection_manager`.
+//
+// This controls what protocol(s) should be used for upstream and how said protocol(s) are configured.
+//
+// This replaces the prior pattern of explicit protocol configuration directly
+// in the cluster. So a configuration like this, explicitly configuring the use of HTTP/2 upstream:
+//
+// .. code::
+//
+//   clusters:
+//     - name: some_service
+//       connect_timeout: 5s
+//       upstream_http_protocol_options:
+//         auto_sni: true
+//       common_http_protocol_options:
+//         idle_timeout: 1s
+//       http2_protocol_options:
+//         max_concurrent_streams: 100
+//        .... [further cluster config]
+//
+// Would now look like this:
+//
+// .. code::
+//
+//   clusters:
+//     - name: some_service
+//       connect_timeout: 5s
+//       typed_extension_protocol_options:
+//         envoy.filters.network.http_connection_manager:
+//           "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpProtocolOptions
```

Comment:
Actually, maybe I'm still a bit confused. We're not specifying HCM as an upstream netwrok filter, but instead as a typed extension protocol options.. that seems a bit weird.

Reply:
I think the goal is to keep protocol specific config (thrift, HTTP, SMTP, etc) from being hard-coded in the cluster config, and instead make cluster config extensible to the particulars of any underlying application layer.
As such this pulls the HTTP config out, and sticks it in a plugin.

Reply:
And as with the prior examples, generally you only need application-specific upstream config if you are doing downstream application-specific config, via a network filter, so the plug-in config gets stored with the network filter.  See RedisProtocolOptions for the template I'm cribbing off of.

Reply:
OK, I think I get it now. I'm curious if it would make sense for HTTP (and Redis) as well, to have a clearer separation of downstream/upstream protocol options. Specifically, we might have the situation where the downstream is neither HTTP or Redis, let's say it's FooRPC network filter, but FooRPC makes HTTP and/or Redis requests via clusters configured with the protocol options for HTTP/Redis. I'd argue that since these are properties of the cluster manager (or connection pool more specifically), this would be a more sensible logical organization.

Specifically tagging @mattklein123 and @markdroth (gRPC implications) for further discussion on this thread.

Reply:
I actually think the use of typed extension here works really well for gRPC, as they can customize with their own alternative. The main open question to me is whether we associate options with downstream filter or with something on the cluster/upstream side.

Reply:
I won't make code/config changes until there's consensus but can you suggest where you think the best place in the API hierarchy might be? I assume somewhere new?

Reply:
Yeah all things being equal I would probably move these protos somewhere else I agree. @alyssawilk didn't you add config for the pluggable upstream HTTP pools? Can it go somewhere near there? Or a new location would be fine with me.

Reply:
yeah, that's api/envoy/extensions/upstreams/ 
We could drop this in api/envoy/extensions/upstreams/http though it's mildly regrettable naming as I think all of this can apply to sidecar style RPCs, and anywhere else we're doing client side HTTP.

Reply:
IMO "upstream" is still used for side-calls, so seems fine to me? Feel free to suggest a new location also if you like.

Reply:
SGTM. I'll wait for at least Harvey or Mark to agree before moving things because it's a lot of moving things.

Reply:
SGTM as well, this makes a lot of sense.

Reply:
In terms of the "key" to use in typed_protocol_extensions directory, I'd prefer we used for the full type name rather than a made-up key as with the existing Redis example. I think you don't need to implement a ton to make this work, maybe just make the key constant the type name for now and file a tracking issue to automagically map from type name to proto later on.

Reply:
yeah upstreams sounds good to me too

Reply:
Ok, so just to check, if we left it in the current location the key would be the string literal

type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpProtocolOptions

and I'll adjust for the new location?

Reply:
Sorry what do you mean above? I don't see upstream in there?

Reply:
was just trying to get clarity if the key is
type.googleapis.com/proto_name
or 
proto_name

Reply:
Oh that I'm not sure. @htuch @lizan?

Reply:
I would make it `envoy.extensions.filters.network.http_connection_manager.v3.HttpProtocolOptions`

We only need this type.googlesapis.com prefix for `Any` fields which force that, otherwise I think we prefer the more compact fully-qualified type name.